### PR TITLE
Update lotus to use v23 proofs

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

File: chain/sync.go [hunk 1]
```diff
@@ -687,14 +694,44 @@ func (syncer *Syncer) VerifyElectionPoStProof(ctx context.Context, h *types.Bloc
...
+		candidates = append(candidates, abi.PoStCandidate{
+			RegisteredProof: rt,
+			PartialTicket:   c.Partial,
+			SectorID:        abi.SectorID{Number: c.SectorID}, // this should not be an ID, we already know who the miner is...
+			ChallengeIndex:  int64(c.ChallengeIndex),
+		})
+	}
+
+	// TODO: why do we need this here?
+	challengeCount := sectorbuilder.ElectionPostChallengeCount(uint64(len(sectorInfo)), 0)
```
Comment (magik6k) [inline chain/sync.go]: Isn't the challengeCount here actually just `len(sectorInfo)`?

File: node/modules/client.go [hunk 2]
```diff
@@ -117,8 +118,9 @@ func StorageClient(h host.Host, ibs dtypes.ClientBlockstore, r repo.LockedRepo,
 }
 
 // RetrievalClient creates a new retrieval client attached to the client blockstore
-func RetrievalClient(h host.Host, bs dtypes.ClientBlockstore, pmgr *paychmgr.Manager, payapi payapi.PaychAPI, resolver retrievalmarket.PeerResolver) retrievalmarket.RetrievalClient {
+func RetrievalClient(h host.Host, bs dtypes.ClientBlockstore, pmgr *paychmgr.Manager, payapi payapi.PaychAPI, resolver retrievalmarket.PeerResolver, ds dtypes.MetadataDS) (retrievalmarket.RetrievalClient, error) {
 	adapter := retrievaladapter.NewRetrievalClientNode(pmgr, payapi)
 	network := rmnet.NewFromLibp2pHost(h)
-	return retrievalimpl.NewClient(network, bs, adapter, resolver)
+	sc := storedcounter.New(ds, datastore.NewKey("/retr"))
```
Comment (magik6k) [inline node/modules/client.go]: check if actually needed

File: chain/sync.go (hunk 1 above)
Comment (whyrusleeping) [reply chain/sync.go]: No, I think that this is right, but i'm confused as to why I need to care here now

File: chain/gen/gen.go [hunk 3]
```diff
@@ -567,23 +572,21 @@ func IsRoundWinner(ctx context.Context, ts *types.TipSet, round int64, miner add
 		return nil, nil
 	}
 
-	var sinfos []ffi.PublicSectorInfo
+	if pset[0].Info.Info.RegisteredProof == 0 {
```
Comment (whyrusleeping) [inline chain/gen/gen.go]: cleaning this up now

File: chain/gen/gen.go [hunk 4]
```diff
@@ -567,23 +572,21 @@ func IsRoundWinner(ctx context.Context, ts *types.TipSet, round int64, miner add
...
+		panic("registered proof coming from proving set is 0")
```
Comment (magik6k) [inline chain/gen/gen.go]: Return an error

File: chain/market/fundmgr.go [hunk 5]
```diff
@@ -31,7 +34,8 @@ func NewFundMgr(sm *stmgr.StateManager, mpool full.MpoolAPI) *FundMgr {
 	}
 }
 
-func (fm *FundMgr) EnsureAvailable(ctx context.Context, addr address.Address, amt types.BigInt) error {
+func (fm *FundMgr) EnsureAvailable(ctx context.Context, addr, wallet address.Address, amt types.BigInt) error {
+	log.Error("ensure available: ", addr, amt)
```
Comment (magik6k) [inline chain/market/fundmgr.go]:
```suggestion
	log.Info("ensure available: ", addr, amt)
```

File: chain/market/fundmgr.go [hunk 6]
```diff
@@ -77,6 +81,7 @@ func (fm *FundMgr) EnsureAvailable(ctx context.Context, addr address.Address, am
 
 	_, r, err := fm.sm.WaitForMessage(ctx, smsg.Cid())
 	if err != nil {
+		log.Error("waiting for message: ", err)
```
Comment (magik6k) [inline chain/market/fundmgr.go]:
```suggestion
		log.Info("waiting for message: ", err)
```

File: markets/storageadapter/client.go [hunk 7]
```diff
@@ -276,29 +276,51 @@ func (c *ClientNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider
...
+					sectorNumber = params.SectorNumber
+					sectorFound = true
+					return false, nil
+				}
+			}
+
 			return false, nil
-		}
+		case builtin.MethodsMiner.ProveCommitSector:
+			log.Errorf("Look! a prove commit!")
```
Comment (magik6k) [inline markets/storageadapter/client.go]:
```suggestion
			log.Infof("Look! a prove commit!")
```
(Or remove)

File: markets/storageadapter/client.go [hunk 8]
```diff
@@ -276,29 +276,51 @@ func (c *ClientNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider
...
+		switch msg.Method {
+		case builtin.MethodsMiner.PreCommitSector:
+			var params miner.SectorPreCommitInfo
+			if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
+				return false, xerrors.Errorf("unmarshal pre commit: %w", err)
+			}
+
+			for _, did := range params.DealIDs {
+				if did == abi.DealID(dealId) {
+					log.Error("FOUND OUR SECTOR!", params.SectorNumber)
```
Comment (magik6k) [inline markets/storageadapter/client.go]:
```suggestion
					log.Info("FOUND OUR SECTOR!", params.SectorNumber)
```

File: markets/storageadapter/client.go [hunk 9]
```diff
@@ -276,29 +276,51 @@ func (c *ClientNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider
...
-			if uint64(dealID) == dealId {
-				found = true
-				break
+			if params.SectorNumber != sectorNumber {
+				return false, nil
 			}
-		}
 
-		return found, nil
+			log.Warn("getting out of here")
```
Comment (magik6k) [inline markets/storageadapter/client.go]:
```suggestion
```

File: markets/storageadapter/client.go [hunk 10]
```diff
@@ -276,29 +276,51 @@ func (c *ClientNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider
...
+
 			return false, nil
-		}
+		case builtin.MethodsMiner.ProveCommitSector:
+			log.Errorf("Look! a prove commit!")
+			var params miner.ProveCommitSectorParams
+			if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
+				return false, xerrors.Errorf("failed to unmarshal prove commit sector params: %w", err)
+			}
+			log.Warn("prove commit: ", params.SectorNumber, sectorFound, sectorNumber)
```
Comment (magik6k) [inline markets/storageadapter/client.go]:
```suggestion
			log.Info("prove commit: ", params.SectorNumber, sectorFound, sectorNumber)
```

File: markets/storageadapter/provider.go [hunk 11]
```diff
@@ -219,6 +219,7 @@ func (n *ProviderNodeAdapter) LocatePieceForDealWithinSector(ctx context.Context
 }
 
 func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provider address.Address, dealID uint64, cb storagemarket.DealSectorCommittedCallback) error {
+	log.Errorf("ON DEAL SECTOR COMMITTED", provider, dealID)
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
	log.Info("ON DEAL SECTOR COMMITTED", provider, dealID)
```

File: markets/storageadapter/provider.go [hunk 12]
```diff
@@ -236,6 +237,7 @@ func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provide
 	}
 
 	called := func(msg *types.Message, rec *types.MessageReceipt, ts *types.TipSet, curH abi.ChainEpoch) (more bool, err error) {
+		log.Errorf("CAllled callllllback")
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
```

File: markets/storageadapter/provider.go [hunk 13]
```diff
@@ -269,29 +271,53 @@ func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provide
...
+		switch msg.Method {
+		case builtin.MethodsMiner.PreCommitSector:
+			var params miner.SectorPreCommitInfo
+			if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
+				return false, xerrors.Errorf("unmarshal pre commit: %w", err)
+			}
+
+			for _, did := range params.DealIDs {
+				if did == abi.DealID(dealID) {
+					log.Error("FOUND OUR SECTOR!", params.SectorNumber)
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
					log.Info("FOUND OUR SECTOR!", params.SectorNumber)
```

File: markets/storageadapter/provider.go [hunk 14]
```diff
@@ -269,29 +271,53 @@ func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provide
...
+					sectorNumber = params.SectorNumber
+					sectorFound = true
+					return false, nil
+				}
+			}
+
 			return false, nil
-		}
+		case builtin.MethodsMiner.ProveCommitSector:
+			log.Errorf("Look! a prove commit!")
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
```

File: markets/storageadapter/provider.go [hunk 15]
```diff
@@ -269,29 +271,53 @@ func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provide
...
+			var params miner.ProveCommitSectorParams
+			if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
+				return false, xerrors.Errorf("failed to unmarshal prove commit sector params: %w", err)
+			}
+			log.Warn("prove commit: ", params.SectorNumber, sectorFound, sectorNumber)
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
			log.Info("prove commit: ", params.SectorNumber, sectorFound, sectorNumber)
```

File: markets/storageadapter/provider.go [hunk 16]
```diff
@@ -269,29 +271,53 @@ func (n *ProviderNodeAdapter) OnDealSectorCommitted(ctx context.Context, provide
...
-		var found bool
-		for _, did := range params.DealIDs {
-			if did == abi.DealID(dealID) {
-				found = true
-				break
+			if params.SectorNumber != sectorNumber {
+				return false, nil
 			}
+
+			log.Warn("getting out of here")
```
Comment (magik6k) [inline markets/storageadapter/provider.go]:
```suggestion
```

**Pull request merged**